src/mpid/ch4/generic/am/mpidig_am_gpu.h

```diff
@@ -0,0 +1,181 @@
+    goto fn_exit;
+#else
+    MPIR_Assert(0);
+    return mpi_errno;
+#endif /* MPIDI_CH4_DIRECT_NETMOD */
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDIG_gpu_lmt_send(const void *buf, MPI_Aint count,
+                                                 MPI_Datatype datatype, MPI_Aint data_sz, int rank,
+                                                 int tag, MPIR_Comm * comm, int context_offset,
+                                                 MPIDI_av_entry_t * addr, MPIR_Request ** request)
+{
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+    if (MPIDI_av_is_local(addr)) {
+        return MPIDIG_gpu_ipc_send_req(buf, count, datatype, data_sz, rank, tag, comm,
+                                       context_offset, NULL, request);
+    }
+    return MPI_ERR_OTHER;
+#else
+    MPIR_Assert(0);
+    return MPI_SUCCESS;
+#endif /* MPIDI_CH4_DIRECT_NETMOD */
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDIG_gpu_ipc_do_put(MPIR_Request * req)
+{
+    int mpi_errno = MPI_SUCCESS;
+    void *dst_buf;
+    MPL_gpu_ipc_open_mem_handle(&dst_buf, MPIDIG_REQUEST(req, req->ireq).mem_handle);
+
+    MPIR_Request *lreq = MPIDIG_REQUEST(req, req->ireq).lreq_ptr;
+    const void *src_buf = MPIDIG_REQUEST(lreq, req->lreq).src_buf;
+    int count = MPIDIG_REQUEST(lreq, req->lreq).count;
+    MPI_Datatype datatype = MPIDIG_REQUEST(lreq, req->lreq).datatype;
+
+    size_t data_sz;
+    MPIDI_Datatype_check_size(datatype, count, data_sz);
+
+    yaksa_request_t yaksa_request;
+    uintptr_t actual_pack_bytes;
+    mpi_errno =
+        yaksa_ipack(src_buf, data_sz, YAKSA_TYPE__BYTE, 0, dst_buf,
+                    data_sz, &actual_pack_bytes, &yaksa_request);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    mpi_errno = yaksa_request_wait(yaksa_request);
+
+  fn_exit:
+    MPL_gpu_ipc_close_mem_handle(dst_buf);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDIG_gpu_ipc_do_get(MPIR_Request * req)
+{
+    int mpi_errno = MPI_SUCCESS;
+    void *src_buf;
+    MPL_gpu_ipc_open_mem_handle(&src_buf, MPIDIG_REQUEST(req, req->ireq).mem_handle);
+
+    MPI_Datatype dst_type = MPIDIG_REQUEST(req, datatype);
+    int count = MPIDIG_REQUEST(req, count);
+    void *pack_buffer = MPIDIG_REQUEST(req, buffer);
+
+    MPL_pointer_type_t mem_type;
+    MPL_gpu_query_pointer_type(pack_buffer, &mem_type);
+
+    int dt_contig;
+    size_t data_sz;
+    MPIDI_Datatype_check_contig_size(dst_type, count, dt_contig, data_sz);
+    if (!dt_contig) {
+        if (mem_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_malloc(&MPIDIG_REQUEST(req, req->ireq).pack_buffer, data_sz);
+            pack_buffer = MPIDIG_REQUEST(req, req->ireq).pack_buffer;
+            MPIDIG_REQUEST(req, req->ireq).needs_unpack = true;
+        } else {
+            MPIDIG_REQUEST(req, req->ireq).pack_buffer = MPL_malloc(data_sz, MPL_MEM_BUFFER);
+            pack_buffer = MPIDIG_REQUEST(req, req->ireq).pack_buffer;
+            MPIDIG_REQUEST(req, req->ireq).needs_unpack = true;
+        }
+    }
```

Comment:
FIXME: Since the sender always exposes the memory handle of the pack buffer this part should be removed.